nucleus/deploy/model_endpoint.py

```diff
@@ -1,12 +1,41 @@
 import concurrent.futures
 from collections import Counter
-from typing import Dict, Optional, Sequence
+from typing import Any, Dict, Optional, Sequence
 
 TASK_PENDING_STATE = "PENDING"
 TASK_SUCCESS_STATE = "SUCCESS"
 TASK_FAILURE_STATE = "FAILURE"
 
 
+class EndpointRequest:
+    """
+    Represents a single request to either a SyncModelEndpoint or AsyncModelEndpoint.
+    Parameters:
+        url: A url to some file that can be read in to a ModelBundle's predict function. Can be an image, raw text, etc.
+        args: A Dictionary with arguments to a ModelBundle's predict function. If the predict function has signature
+            predict_fn(foo, bar), then the keys in the dictionary should be 'foo' and 'bar'. Values must be native Python
+            objects.
+        return_pickled: Whether the output should be a pickled python object, or directly returned serialized json
+        request_id: A user-specifiable id for requests. Should be unique.
```

Comment:
When/why would a user specify this? I was imagining this would be populated by us server-side.

Reply:
The thing this `request_id` is trying to solve is that a user needs to be able to associate model inference results with a given model input. For example, for the nucleus integration we need to maintain some map from s3URL of input images to task results, so this request_id can be set to that s3URL. Before, I was implicitly using the url as the request_id, but we don't necessarily have urls anymore.

I suppose that we could just keep the association implicitly in the order of some list somewhere but then the user would have to do that too.

Reply:
Ok it sounds like you're basically using this as a `reference_id` then? Can we just have our client generate a UUID? Then the nucleus integration (which one day will be server-side) can just maintain that `Request` -> `Response` mapping, with the hash key being this UUID.

Reply:
I left in the ability for clients to specify their own `request_id`s, but otherwise the client will generate a uuid for the `request_id`

Reply:
I would err on the more restrictive side and not allow `request_id`s, because I'm imagining a hapless user passing in "42" hardcoded and never changing it.

Reply:
(probably should have specified this but) The uniqueness condition is actually more permissive than "globally unique", but instead is "the EndpointRequests producing a given BatchResponse must have distinct `request_ids`". I think that mostly gets rid of the hardcoding problem? (i.e. it doesn't matter if the user runs the script twice with the same value of `request_id`, but if the user hardcodes a value for `request_id` somewhere and copies the code that would be a problem)

I do think we should allow users to set `request_id`s though, since otherwise it's much more annoying for users to keep track of an association from `EndpointRequest`/`EndpointResponse` to something a user cares about, e.g. if the user wants to run inference on N images, they need to keep track of which `EndpointResponse` gets associated with which image. 

With user-specified `request_id`s the user can do `endpoint_response_for_image_i = job.responses[request_id_for_image_i]`; with non-user-specified `request_ids` the user would need to do 
```
# Go through the list of EndpointRequests to find the request_id corresponding to a given image, comparing either `url` or `args`. Comparing values seems reasonable in the `url` case but ugly in the `args` case.
# job.responses[the request_id we found]
```
or use the ordering of `EndpointRequest` fed in and the corresponding ordering of `EndpointResponses` (currently `EndpointResponses` are returned in a dictionary though)
(or something I haven't thought of)

One simple thing we could do to catch errors earlier is just to check that the user passes a set of distinct `request_id`s to the batch_predict function at the very least

Reply:
Oh I see. I think this is happening because the client library is still operating under the paradigm where batch jobs are tied to endpoints instead of being standalone. In the eventual state, I would expect the batch request to be its own class (not a `predict_batch` method off of an endpoint class), and result would be a url to a file/directory that contains outputs, with the relative order not being guaranteed anyway. In other words, this is needed only because have to support the current janky client-side integration. In that case I guess this is fine, but prior to launch I definitely would want `request_id` to be gone altogether from the client, so please add a TODO.

I was tempted to suggest that we call this `reference_id`, but that's misleading in a _different_ way (Nucleus actually persists those, so for us to use them as ephemeral client-side IDs could be confusing due to inconsistency).

Reply:
Yeah it's pretty much there because of the async batch class